src/main/java/org/opentripplanner/graph_builder/module/osm/naming/SidewalkNamer.java

```diff
@@ -0,0 +1,304 @@
+ * intersection the sidewalk is also split. It doesn't work well for sidewalks that go around
+ * the corner, like https://www.openstreetmap.org/way/1059101564. These cases are, however, detected
+ * by the above algorithm and the sidewalk name remains the same.
+ */
+public class SidewalkNamer implements EdgeNamer {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SidewalkNamer.class);
+  private static final int MAX_DISTANCE_TO_SIDEWALK = 50;
+  private static final double MIN_PERCENT_IN_BUFFER = .85;
+  private static final int BUFFER_METERS = 25;
+
+  private HashGridSpatialIndex<EdgeOnLevel> streetEdges = new HashGridSpatialIndex<>();
+  private Collection<EdgeOnLevel> unnamedSidewalks = new ArrayList<>();
+  private PreciseBuffer preciseBuffer;
+
+  @Override
+  public I18NString name(OSMWithTags way) {
+    return way.getAssumedName();
+  }
+
+  @Override
+  public void recordEdges(OSMWithTags way, StreetEdgePair pair) {
+    if (way.isSidewalk() && way.needsFallbackName() && !way.isExplicitlyUnnamed()) {
+      pair
+        .asIterable()
+        .forEach(edge -> unnamedSidewalks.add(new EdgeOnLevel(edge, way.getLevels())));
+    }
+    if (way.isNamed() && !way.isLink()) {
+      // we generate two edges for each osm way: one there and one back. since we don't do any routing
+      // in this class we don't need the two directions and keep only one of them.
+      var edge = pair.pickAny();
+      streetEdges.insert(
+        edge.getGeometry().getEnvelopeInternal(),
+        new EdgeOnLevel(edge, way.getLevels())
+      );
+    }
+  }
+
+  @Override
+  public void postprocess() {
+    ProgressTracker progress = ProgressTracker.track(
+      "Assigning names to sidewalks",
+      500,
+      unnamedSidewalks.size()
+    );
+
+    this.preciseBuffer = new PreciseBuffer(computeCentroid(), BUFFER_METERS);
+
+    final AtomicInteger namesApplied = new AtomicInteger(0);
+    unnamedSidewalks
+      .parallelStream()
+      .forEach(sidewalkOnLevel -> {
+        assignNameToSidewalk(sidewalkOnLevel, namesApplied);
+
+        // Keep lambda! A method-ref would cause incorrect class and line number to be logged
+        // noinspection Convert2MethodRef
+        progress.step(m -> LOG.info(m));
+      });
+
+    LOG.info(
+      "Assigned names to {} of {} of sidewalks ({}%)",
+      namesApplied.get(),
+      unnamedSidewalks.size(),
+      DoubleUtils.roundTo2Decimals((double) namesApplied.get() / unnamedSidewalks.size() * 100)
+    );
+
+    LOG.info(progress.completeMessage());
+
+    // set the indices to null so they can be garbage-collected
+    streetEdges = null;
+    unnamedSidewalks = null;
+  }
+
+  /**
+   * Compute the centroid of all sidewalk edges.
+   */
+  private Coordinate computeCentroid() {
+    var envelope = new Envelope();
+    unnamedSidewalks.forEach(e ->
+      envelope.expandToInclude(e.edge.getGeometry().getEnvelopeInternal())
```

Comment:
This is unpacking a lot of geometries and making a lot of small objects to establish their envelopes. Just making an envelope around the intersection endpoints would be a lightweight approach yielding a very similar result. Not sure if you care about the optimization, just pointing it out and it could become significant on larger graphs.

Reply:
I've reworked this in https://github.com/opentripplanner/OpenTripPlanner/commit/f995865fbe33c64691c5ea91f3f546057facc60b